File: common/common-web/src/main/java/io/vanaheimr/common/web/RequestContextFilter.java
package io.vanaheimr.common.web;

import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.filter.FilterChain;
import io.micronaut.http.filter.HttpFilter;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;

@Singleton
public class RequestContextFilter implements HttpFilter {
    @Override
    public Publisher<? extends HttpResponse<?>> doFilter(HttpRequest<?> request, FilterChain chain) {
        return chain.proceed(request);
    }
}

File: common/common-web/src/main/kotlin/io/vanaheimr/common/web/RequestContextFilter.kt
package io.vanaheimr.common.web

import io.micronaut.core.order.Ordered
import io.micronaut.http.HttpRequest
import io.micronaut.http.MutableHttpResponse
import io.micronaut.http.annotation.Filter
import io.micronaut.http.filter.HttpServerFilter
import io.micronaut.http.filter.ServerFilterChain
import io.vanaheimr.common.core.RequestContext
import kotlinx.coroutines.reactive.awaitFirstOrNull
import kotlinx.coroutines.reactor.mono
import org.reactivestreams.Publisher

@Filter("/**")
class RequestContextFilter : HttpServerFilter {

    override fun getOrder() = Ordered.HIGHEST_PRECEDENCE

    override fun doFilter(request: HttpRequest<*>, chain: ServerFilterChain): Publisher<MutableHttpResponse<*>> {
        return mono(RequestContext()) {
            chain.proceed(request).awaitFirstOrNull()
        }
    }
}